File: targets/java-web/src/main/java/com/example/vulnapp/vulns/Vuln34Cors.java
package com.example.vulnapp.vulns;

import org.springframework.web.bind.annotation.*;
import javax.servlet.http.HttpServletResponse;

@RestController
public class Vuln34Cors {
    @GetMapping("/vuln34/data")
    public String data(@RequestHeader(value = "Origin", required = false) String origin, HttpServletResponse response) {
        if (origin != null && !origin.isEmpty()) {
            // VULN:VULN-34:CWE-942:config arbitrary Origin reflected + credentials allowed
            response.setHeader("Access-Control-Allow-Origin", origin);
            response.setHeader("Access-Control-Allow-Credentials", "true");
        }
        return "{\"secret\":\"account-balance-42000\"}";
    }
}

File: targets/kotlin-web/src/main/kotlin/com/example/vulnapp/vulns/Vuln34Cors.kt
package com.example.vulnapp.vulns

import org.springframework.web.bind.annotation.*


@RestController
class Vuln34Cors {

    @GetMapping("/vuln34/data")
    fun data(@RequestHeader(value = "Origin", required = false) origin: String?,
             response: javax.servlet.http.HttpServletResponse): String {
        if (!origin.isNullOrEmpty()) {
            // VULN:VULN-34:CWE-942:config arbitrary Origin reflected + credentials allowed
            response.setHeader("Access-Control-Allow-Origin", origin)
            response.setHeader("Access-Control-Allow-Credentials", "true")
        }
        return "{\"secret\":\"account-balance-42000\"}"
    }
}
